OPA.ps1

```diff
@@ -13,50 +13,111 @@ $DebugPreference = "Continue"
 $InformationPreference = "Continue"
 $ErrorActionPreference = "Stop"
 
-# Set expected version and OutFile path
-$ExpectedVersion = "0.42.1"
-$OPAExe = "opa_windows_amd64.exe"
-$InstallUrl = "https://openpolicyagent.org/downloads/v$($ExpectedVersion)/$OPAExe"
-$OutFile=(Join-Path (Get-Location).Path $InstallUrl.SubString($InstallUrl.LastIndexOf('/')))
-$ExpectedHash ="5D71028FED935DC98B9D69369D42D2C03CE84A7720D61ED777E10AAE7528F399"
-
 # Download files
-try {
-    Write-Information "Downloading $InstallUrl"
-    $WebClient = New-Object System.Net.WebClient
-    $WebClient.DownloadFile($InstallUrl, $OutFile)
-    Write-Information ""
-    Write-Information "`nDownload of `"$OutFile`" finished."
+function Get-OPAExecutable {
+    param (
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('version')]
+        [string]$ExpectedVersion,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('opa')]
+        [string]$OPAExe
+    )
+
+    $InstallUrl = "https://openpolicyagent.org/downloads/v$($ExpectedVersion)/$OPAExe"
+    $OutFile=(Join-Path (Get-Location).Path $InstallUrl.SubString($InstallUrl.LastIndexOf('/')))
+
+    try {
+        $WebClient = New-Object System.Net.WebClient
+        $WebClient.DownloadFile($InstallUrl, $OutFile)
+        Write-Information -MessageData "Installed the latest acceptable version of ${OPAExe} version ${ExpectedVersion}." | Out-Host
+    }
+    catch {
+        $Error[0] | Format-List -Property * -Force | Out-Host
+        Write-Error "Unable to download OPA executable. To try manually downloading, see details in README under 'Download the required OPA executable'" | Out-Host
+    }
+    finally {
+        $WebClient.Dispose()
+    }
 }
-catch {
-    Write-Error "An error has occurred: Unable to download OPA executable. To try manually downloading, see details in README under 'Download the required OPA executable'"
+
+
+function Confirm-OPAExeIsValid {
+    param (
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('opa')]
+        [string]$OPAExe,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('version')]
+        [string]$ExpectedVersion
+    )
+
+    # Hash checks
+    $ExpectedHash ="5D71028FED935DC98B9D69369D42D2C03CE84A7720D61ED777E10AAE7528F399"
```

Comment:
Consider adding function to always download hash for the given opa version. It is low cost to download hash conpared to downloading full executable.

Reply:
Added the functionality